File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/common/Transformer.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.ml.common

import org.apache.flink.api.scala.DataSet

/** Base trait for an algorithm which transforms the input data to some output data.
  *
  * A [[Transformer]] is used to transform input data to some output data. Transformations might
  * be feature extractors, feature mappings, whitening or centralization just to name a few.
  *
  * [[Transformer]] can be chained with other [[Transformer]] creating a [[ChainedTransformer]],
  * which again can be chained. Chaining a [[Transformer]] with a [[Learner]] creates a
  * [[ChainedLearner]] which terminates a pipeline.
  *
  * A [[Transformer]] implementation has to implement the method `transform`, which defines how
  * the input data is transformed into the output data.
  *
  * @tparam IN Type of incoming elements
  * @tparam OUT Type of outgoing elements
  */
trait Transformer[IN, OUT] extends WithParameters {
  def chain[CHAINED](transformer: Transformer[OUT, CHAINED]):
  ChainedTransformer[IN, OUT, CHAINED] = {
    new ChainedTransformer[IN, OUT, CHAINED](this, transformer)
  }

  def chain[CHAINED](learner: Learner[OUT, CHAINED]): ChainedLearner[IN, OUT, CHAINED] = {
    new ChainedLearner[IN, OUT, CHAINED](this, learner)
  }

  def transform(input: DataSet[IN], transformParameters: ParameterMap = ParameterMap.Empty):
  DataSet[OUT]
}

File: flink-ml-parent/flink-ml-api/src/main/java/org/apache/flink/ml/api/core/Transformer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.ml.api.core;

import org.apache.flink.annotation.PublicEvolving;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableEnvironment;

/**
 * A transformer is a {@link PipelineStage} that transforms an input {@link Table} to a result
 * {@link Table}.
 *
 * @param <T> The class type of the Transformer implementation itself, used by {@link
 *            org.apache.flink.ml.api.misc.param.WithParams}
 */
@PublicEvolving
public interface Transformer<T extends Transformer<T>> extends PipelineStage<T> {
	/**
	 * Applies the transformer on the input table, and returns the result table.
	 *
	 * @param tEnv  the table environment to which the input table is bound.
	 * @param input the table to be transformed
	 * @return the transformed table
	 */
	Table transform(TableEnvironment tEnv, Table input);
}
